### PR TITLE
Avoid lenient Number cast in writeNativeValue

### LINKED ISSUE #8332: Avoid lenient Number cast in writeNativeValue
In the past a conversion through `Number` "interface" has proven to be able to mask bugs in the code.
We should remove these conversions from `io.trino.spi.type.TypeUtils#writeNativeValue`.

### PR DESCRIPTION
**Pull request by pvliborio**
Fixes #8332

- Remove `Number.class` conversion from `io.trino.spi.type.TypeUtils#writeNativeValue`

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/kafka/TestKafkaAvroWritesSmokeTest.java [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 import static io.trino.tempto.assertions.QueryAssert.Row.row;
 import static io.trino.tempto.assertions.QueryAssert.assertThat;
-import static io.trino.tempto.assertions.QueryAssert.assertQueryFailure
+import static io.trino.tempto.assertions.QueryAssert.assertQueryFailure;
```
Comment (Praveen2112) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/kafka/TestKafkaAvroWritesSmokeTest.java:21]: Can you squash it with the previous commit ?

**Review by Praveen2112 — COMMENTED** [review]
Can you please rebase it.

**Comment by findepi** [comment]
@pvliborio thank you for your PR!

can you please separate `Migrate from assertThat to assertQueryFailure` and `Remove Number from TypeUtils` into separate PRs?
the two are not related. This will help us with the review process.
Thanks

**Comment by findepi** [comment]
Oh, i see https://github.com/trinodb/trino/pull/8342 now. So please just remove the `Migrate from assertThat to assertQueryFailure` commit from here.

**Comment by Praveen2112** [comment]
Thanks for working on this. Can we have the commit message similar to `Avoid lenient Number cast in writeNativeValue` i.e Issue title itself. I guess you have signed the CLA.

**Comment by pvliborio** [comment]
> Thanks for working on this. Can we have the commit message similar to `Avoid lenient Number cast in writeNativeValue` i.e Issue title itself. I guess you have signed the CLA.

I signed. Also, revised the code after looking the CI errors. Thanks!

File: core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java [hunk 2]
```diff
@@ -73,10 +73,18 @@ else if (type.getJavaType() == boolean.class) {
             type.writeBoolean(blockBuilder, (Boolean) value);
         }
         else if (type.getJavaType() == double.class) {
-            type.writeDouble(blockBuilder, ((Number) value).doubleValue());
+            if(value instanceof Float) {
+                type.writeDouble(blockBuilder, ((Float) value).doubleValue());
+            } else {
+                type.writeDouble(blockBuilder, ((Double) value).doubleValue());
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:79]: if we were writing this code anew, we  would require value to be Double.

what would it take to have that?

File: core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java [hunk 3]
```diff
@@ -73,10 +73,18 @@ else if (type.getJavaType() == boolean.class) {
...
+            }
         }
         else if (type.getJavaType() == long.class) {
-            type.writeLong(blockBuilder, ((Number) value).longValue());
+            if(value instanceof Integer) {
+                type.writeLong(blockBuilder, ((Integer) value).longValue());
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:84]: as above, it would be more natural to require Long here.

File: core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java (hunk 2 above)
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:79]: to be clear: changing from "Number" to "Float or Double" is an important improvement

Comment (pvliborio) [reply core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:79]: Hi @findepi, I didn't understand what I need to change. There was an error in the CI related to the cast of an Integer as an Long, so I verified if the instance of the value was Integer and made the cast using longValue(). Same to Double and Float.

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:79]: > There was an error in the CI related to the cast of an Integer as an Long,

do you remember what was failing?

Comment (pvliborio) [reply core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:79]: > > There was an error in the CI related to the cast of an Integer as an Long,
> 
> do you remember what was failing?

java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')

The same is appearing now with Short and Long.

Comment (hashhar) [reply core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:79]: I think there are a lot of callers which send wider types than required. Those will need to be changed to just send Double instead of Float or something else.

If you can try to figure out how many callers you need to fix then we can decide to this now or as a follow-up. I would prefer now unless it seems like a very large change.

Comment (pvliborio) [reply core/trino-spi/src/main/java/io/trino/spi/type/TypeUtils.java:79]: > I think there are a lot of callers which send wider types than required. Those will need to be changed to just send Double instead of Float or something else.
> 
> If you can try to figure out how many callers you need to fix then we can decide to this now or as a follow-up. I would prefer now unless it seems like a very large change.

I couldn't find out, there are many callers for this function and for the calling functions. I thought of the following:
```suggestion
public static void writeNativeValue(Type type, BlockBuilder blockBuilder, @Nullable Object value)
    {
        ...
        else if (type.getJavaType() == double.class && value instanceof Double) {
            type.writeDouble(blockBuilder, ((Double) value).doubleValue());
        }
        else if (type.getJavaType() == long.class && value instanceof Long) {
            type.writeLong(blockBuilder, ((Long) value).longValue());
        }
        else if (type.getJavaType() == Slice.class) {
        ...
        else {
            type.writeObject(blockBuilder, value);
        }
    }
```

Would that be a good resolution, even a temporary one?

**Comment by pvliborio** [comment]
@hashhar I tried the solution I commented and not worked. What would be the problem of casting an Integer to a Long? Because a Long is just an 8-byte integer.

**Comment by hashhar** [comment]
`type.writeLong(blockBuilder, ((Long) value).longValue());`

This doesn't work when `value` is an `Integer` because the type of `value` is an `Object`. So you think you know it's an Integer but the JVM doesn't. So you need two casts.

`type.writeLong(blockBuilder, ((Integer) value).longValue());`

Which brings you back to the original problem that you'll need two branches - one for Long `value` and one for `Integer` value. The only proper fix is to change the callers to always pass a `Long`.

You can add something like `verify(value instanceof Long)` in the branch and look at stack trace of the failing tests and adjust the callers accordingly.

**Comment by pvliborio** [comment]
> `type.writeLong(blockBuilder, ((Long) value).longValue());`
> 
> This doesn't work when `value` is an `Integer` because the type of `value` is an `Object`. So you think you know it's an Integer but the JVM doesn't. So you need to casts to `Integer` first.
> 
> `type.writeLong(blockBuilder, ((Integer) value).longValue());`
> 
> Which brings you back to the original problem that you'll need two branches - one for Long `value` and one for `Integer` value. The only proper fix is to change the callers to always pass a `Long`.
> 
> You can add something like `verify(value instanceof Long)` in the branch and look at stack trace of the failing tests and adjust the callers accordingly.

I was trying to run the testes on my local machine but it always fails with a different exceptions. Could you tell me the right command to run at least the trino-main test on the local machine (WSL)? Or I can use the CI?

**Comment by hashhar** [comment]
> I was trying to run the testes on my local machine but it always fails with a different exceptions. Could you tell me the right command to run at least the trino-main test on the local machine (WSL)? Or I can use the CI?

Sorry for the delay. You can run the trino-main tests in IntelliJ by right-clicking on the `trino-main` module and selecting "Run > All Tests".
Screenshot 2021-07-08 at 10 20 14 AM

If you run into issues with that you can use the CLI to do it too as `./mvnw test -pl :trino-main` or if all else fails you can use the CI too.

**Comment by pvliborio** [comment]
@hashhar Hi, I think that is resolved. The failed check isn't related with this issue, in the previous CI the test passed so idk what happened. Can you see it so I can squash the commits if everything looks fine? Thank you a lot for all the help!

**Comment by hashhar** [comment]
Indeed the failed test is a known flaky test - https://github.com/trinodb/trino/issues/8477.

Thanks a lot for working on this @pvliborio - even though it required more effort than what a normal `good first issue` does.

I think it might make sense to squash the commits now for review and apply any [review comments as `fixups`](https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html)

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java [hunk 4]
```diff
@@ -261,7 +261,7 @@ private void addColumnsRecords(QualifiedTablePrefix prefix)
                         tableName.getSchemaName(),
                         tableName.getTableName(),
                         column.getName(),
-                        ordinalPosition,
+                        ((Integer) ordinalPosition).longValue(),
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java:264]:
```suggestion
                        (long) ordinalPosition),
```

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java:264]: however, you can actually declare `ordinalPosition` as  a `long` instead.

File: core/trino-main/src/main/java/io/trino/testing/MaterializedResult.java [hunk 5]
```diff
@@ -253,13 +253,13 @@ else if (BIGINT.equals(type)) {
             type.writeLong(blockBuilder, (Long) value);
         }
         else if (INTEGER.equals(type)) {
-            type.writeLong(blockBuilder, (Integer) value);
+            type.writeLong(blockBuilder, ((Integer) value).longValue());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/testing/MaterializedResult.java:256]:
```suggestion
            type.writeLong(blockBuilder, (Integer) value);
```

-- `.longValue()` is redundant here

(same below)

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/TypeUtils.java [hunk 6]
```diff
@@ -142,14 +142,23 @@ private static Object jdbcObjectToTrinoNative(ConnectorSession session, Object j
...
+        if (TINYINT.equals(type)) {
+            return ((Byte) jdbcObject).longValue();
+        }
+
+        if (SMALLINT.equals(type)) {
+            return ((Short) jdbcObject).longValue();
+        }
+
+        if (INTEGER.equals(type)) {
+            return ((Integer) jdbcObject).longValue();
```
Comment (findepi) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/TypeUtils.java:159]: i would consider writing this as:

```suggestion
            return (long) (int) jdbcObject;
```

-- avoids boxing.

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/TypeUtils.java [hunk 7]
```diff
@@ -164,7 +175,7 @@ private static Object jdbcObjectToTrinoNative(ConnectorSession session, Object j
         }
 
         if (REAL.equals(type)) {
-            return floatToRawIntBits((float) jdbcObject);
+            return ((Integer) floatToRawIntBits((float) jdbcObject)).longValue();
```
Comment (findepi) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/TypeUtils.java:178]: `floatToRawIntBits` returns you an integer, so you need just a cast to long:

```suggestion
            return (long) floatToRawIntBits((float) jdbcObject);
```

-- shorter & no boxing needed

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/TypeUtils.java [hunk 8]
```diff
@@ -142,14 +142,25 @@ private static Object jdbcObjectToTrinoNative(ConnectorSession session, Object j
...
         if (BOOLEAN.equals(type)
-                || TINYINT.equals(type)
-                || SMALLINT.equals(type)
-                || INTEGER.equals(type)
                 || BIGINT.equals(type)
                 || DOUBLE.equals(type)) {
             return jdbcObject;
         }
 
+        System.out.println("Type " + type.getDisplayName() + " - " + type.getJavaType().getName() + " " + type.getClass().getName() + " - min:" + jdbcObject + " (" + (jdbcObject instanceof Integer) + ")");
```
Comment (findepi) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/TypeUtils.java:150]: remove

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 9]
```diff
@@ -312,7 +312,7 @@ else if (fieldValue instanceof Block) {
                     singleRowBlockWriter.appendStructure((Block) fieldValue);
                 }
                 else if (fieldValue instanceof Integer) {
-                    INTEGER.writeLong(singleRowBlockWriter, ((Integer) fieldValue).intValue());
+                    INTEGER.writeLong(singleRowBlockWriter, ((Integer) fieldValue).longValue());
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:315]: this looks like an unrelated change, please revert.

the code here can be simplified, let me do this in https://github.com/trinodb/trino/pull/8576

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
@pvliborio can you please update the commit author to reflect your name or login, but not "root"?

```
commit e53079bfd1844e6ab886f286f284cffde251e060 
Author:     root <liborio@aluno.puc-rio.br>
AuthorDate: Tue Jun 22 01:47:23 2021 -0300
Commit:     root <liborio@aluno.puc-rio.br>
CommitDate: Mon Jul 12 12:00:45 2021 -0300
```

the author  name is typically used in IDEs when annotating the code.

you can update this with:

```
git config --global user.name "Paulo Vítor"
git commit --amend --reset-author
```

**Comment by pvliborio** [comment]
@findepi Done! Thank you!
Thank you too @hashhar !

**Review by findepi — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java [hunk 10]
```diff
@@ -284,6 +284,7 @@ public static Object convert(Object value, Type type)
         if (value == null) {
             return null;
         }
+
```
Comment (hashhar) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java:287]: nit: revert whitespace change

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java [hunk 11]
```diff
@@ -304,6 +305,11 @@ public static Object convert(Object value, Type type)
         if (type instanceof Types.FloatType) {
             return Float.floatToIntBits((Float) value);
         }
+
```
Comment (hashhar) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionTable.java:308]: @pvliborio In a separate commit can you also change:

```
        if (type instanceof Types.TimeType) {
            return ((Long) value) * PICOSECONDS_PER_MICROSECOND;
        }
```

to 

```
        if (type instanceof Types.TimeType) {
            return multiplyExact(value, PICOSECONDS_PER_MICROSECOND);
        }
```

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java [hunk 12]
```diff
@@ -250,7 +250,7 @@ private void addColumnsRecords(QualifiedTablePrefix prefix)
     {
         for (Map.Entry<SchemaTableName, List<ColumnMetadata>> entry : listTableColumns(session, metadata, accessControl, prefix).entrySet()) {
             SchemaTableName tableName = entry.getKey();
-            int ordinalPosition = 1;
+            long ordinalPosition = 1;
```
Comment (hashhar) [inline core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java:253]: Why was this change required?

**Review by hashhar — APPROVED** [review]
Thanks a lot for cleaning this up @pvliborio.

A couple of comments. Looks good otherwise.

File: core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java (hunk 12 above)
Comment (hashhar) [reply core/trino-main/src/main/java/io/trino/connector/informationschema/InformationSchemaPageSource.java:253]: Oh, I understand now. This gets passed on as input to `writeNativeValue`. 👍

**Comment by pvliborio** [comment]
@hashhar I think it's done!

**Comment by hashhar** [comment]
Great work @pvliborio. Thanks for your patience.

Can you please reword the commit messages to something like `Remove lenient Number cast in writeNativeValue` and `Change PartitionTable#convert to throw on overflow in multiplication` (we are explicitly NOT catching the exception so it's more technically correct).

Will merge once CI is done.

**Review by hashhar — APPROVED** [review]
